### PR TITLE
[WIP][PROTOTYPE] Add MetadataLog and HDFSMetadataLog

### PR DESCRIPTION
**Pull request by zsxwing**
## What changes were proposed in this pull request?

- [x] Add MetadataLog and HDFSMetadataLog for metadata reliably storage.
- [x] Update FileStreamSource to use HDFSMetadataLog instead of managing metadata by itself.
- [x] Add FileStreamSink
- [x] Update StreamExecution to use HDFSMetadataLog to manage the progress. Now Sink doesn't need to store its progress.
- [ ] unit tests for MetadataLog and HDFSMetadataLog

## How was this patch tested?

unit tests.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 1]
```diff
@@ -59,6 +59,19 @@ class StreamExecution(
   private val sources =
     logicalPlan.collect { case s: StreamingRelation => s.source }
 
+  private val metadataLog: Option[MetadataLog[Offset]] =
+    try {
+      Some(new HDFSMetadataLog[Offset](sqlContext, sqlContext.getConf("stream.metadata.path")))
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: Please ignore how to get the metadata path now. Will fix it when we have an agreement on this one.

**Comment by zsxwing** [comment]
cc @rxin @tdas @marmbrus

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+import org.apache.hadoop.fs.{Path, PathFilter}
+
+import org.apache.spark.SparkException
+import org.apache.spark.serializer.JavaSerializer
+import org.apache.spark.sql.SQLContext
+import org.apache.spark.util.Utils
+
+/**
+ * An [[MetadataLog]] implementation based on HDFS. [[HDFSMetadataLog]] uses the specified `path`
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala]: Nit `A [[MetadataLog]]`

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala]: We also need to talk about semantics here:
 - what atomicity guarantees are required of the file system (does it work on S3?)?
 - is it corrupted i if there are two writers or does one just get an exception?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+      path.getName.toLong
+      true
+    } catch {
+      case _: NumberFormatException => false
+    }
+  }
+
+  private val serializer = new JavaSerializer(sqlContext.sparkContext.conf).newInstance()
+
+  private def tryAcquireLock(): Unit = {
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala]: Do we need this?  Why isn't a collision on the next file written not sufficient.  If a job dies without deleting, won't this make it impossible to restart?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Sink.scala [hunk 4]
```diff
@@ -20,28 +20,19 @@ package org.apache.spark.sql.execution.streaming
...
-   * transactionally as data is eventually persisted.
+   * Accepts a new batch of data as well as a `batchId` that denotes how far in the input
+   * data computation has progressed to.
+   *
+   * In the case of a failure, the sink may be called with the same batch id multiple times.
+   * The implementation should be idempotent for the same batch id. In other words, it should either
+   * ignore the same batch or override the existing results for the same batch id.
    */
-  def addBatch(batch: Batch): Unit
+  def addBatch(batchId: Long, batch: Batch): Unit
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Sink.scala]: We don't need batches anymore do we?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala (hunk 1 above)
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: maybe toss the query name in here for now?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 5]
```diff
@@ -360,59 +357,6 @@ class FileStreamSourceSuite extends FileStreamSourceTest with SharedSQLContext {
     Utils.deleteRecursively(tmp)
   }
 
-  test("fault tolerance with corrupted metadata file") {
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: Why remove this?

**Comment by marmbrus** [comment]
This is a good start.  Can you remove the sink changes and open this against apache?

**Comment by zsxwing** [comment]
> This is a good start. Can you remove the sink changes and open this against apache?

Do you mean only submit `MetadataLog`, `HDFSMetadataLog` and the changes to `FileStreamSource`?

**Comment by marmbrus** [comment]
Yeah, that part seems self contained and we would need to spend a bunch of time writing tests for the sink before we can submit that PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala (hunk 3 above)
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala]: > If batchId's metatdata has already been stored, this method does nothing.
As the semantics of `add` is `If batchId's metatdata has already been stored, this method does nothing.`, if the writer A sees a file created by the other writer B, it won't write the file. So A won't fail and will use the metadata written by B.

>  If a job dies without deleting, won't this make it impossible to restart?
See my comment in the new Apache PR.

**Pull request closed without merging**